### DISCUSSION TITLE
自定义集合表达式的问题

### DISCUSSION THREAD

**Discussion by tsctsc6** [BYJRK/DotNet-Discussions#307] [C# 技术交流]
根据[官方文档](https://learn.microsoft.com/zh-cn/dotnet/csharp/language-reference/operators/collection-expressions)，尝试自定义集合表达式来构造`RangeCollention`，[源码](https://github.com/tsctsc6/collection-expressions)

在Program.cs 16行，去掉注释会报一个奇怪的错，“无法将Rang隐式变换为int”。不是，谁要他变换为int了？在RangeExtension.cs中，我明明写的是`ReadOnlySpan<Range>`。

![image](https://github.com/user-attachments/assets/2c89d73d-208f-41e6-b68b-602618c13b73)

仔细看了一下，编译器把方括号识别为`Index implicit operator Index(int)`，为何？

**Comment by KiyanYang** [comment]
当前编译器通过`IEnumerator<T> GetEnumerator`来识别类型，对于你的代码会识别到`int`。

可以参考 https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-12.0/collection-expressions#the-notion-of-iteration-type-is-not-applied-consistently-throughout-conversions

**Reply by tsctsc6** [reply]
可是我问的是16行，
`RangeCollention r3 = [0..5, 10..15];`
这一句无法构造类型RangeCollention

**Reply by KiyanYang** [reply]
看 https://github.com/tsctsc6/collection-expressions/blob/main/RangeExtension%2FUtilities%2FRangeExtension.cs#L25-L25 的返回值，编译器会认为你的`RangeCollention`是`int`的迭代类型，所以`0..5`会被编译器尝试转换为`int`，你可以把返回值改成`IEnumerator<long>` 再看看报错信息

编译器不通过继承的接口识别类型，也不通过`CollectionBuilder`来识别类型，而是通过`IEnumerator<T> GetEnumerator()`来识别类型为 T的迭代类型。

**Reply by tsctsc6** [reply]
好，我明白你的意思了

**Comment by tsctsc6** [comment]
也就是说，我想象中的功能就无法实现了？

**Reply by KiyanYang** [reply]
应该是，需要多一些操作

比如你的`new([0..5, 10..15])`，或者扩展方法或[重载操作符](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcAMACOSAsA3OlrkgHQAyAllAI7EY57kBK0w1AtgKbkDCAey4AHagBseYAMpSAbtQDGPAM6MSAZlwJs/bAG902Y9p2G0Jy9lYBDKAHMeQ8ZMWdBUbGDuOV2ALzYANqY5OQArAA02Eih5EjhALqMViYAZoJgPDaKABYAFLTA2NQlngB+3g6qAJRGqVZ4AJyFNSkmAL7oXWjoQc6u7lAAQhASACZS+cAAniI8gmn5ttUDPG7UHjXRUDa8i/n8WTbAPDU1iZra1j5Ogi7rQ8vZ4wDyUOIz0iJ2ADwrjgAfF5bioagZ6sYRGBqHITjwvC8PJ8btUgolsAB9Kq+AIg6oqcgAFUEAEEwN4Zvk2pDsLS4Fo8AA2VGONYbDy6Y6nZ42N4fL4/KD/W7AnG1WmWfzAqA8ADu+XFYPaxnpWgAkgBRKAQXjeYCZEXVYEAcR4wG1uqkJ0y1MlJmlWKV5FJKktepsACNJNTyGaLTqPQawNSVXSLCYGWQWXgNL8isDBPN9ZlsOVlrd2UN8b5weYGtgMsc8vk4WAvGUcxKIwXjEXSzZy6VAmByNJgI3gOQAGo2cQQHiEErYX5ecja8Y9vsDofUADUc7qNdrljwSFwAHYSmGeh0gA===)

**Comment by tsctsc6** [comment]
总之感谢你的解答

**Comment by tsctsc6** [comment]
但是我去试了一下文档的这段代码：
```
using System.Collections;
using System.Collections.Generic;

class MyCollection : IEnumerable
{
    IEnumerator IEnumerable.GetEnumerator() => throw null;

    public IEnumerator<string> GetEnumerator() => throw null; 
    public void Add(object l) => throw null;
}

class Program
{
    static void Main()
    {
        Test("2", 3); // error CS1503: Argument 2: cannot convert from 'int' to 'string'
        Test(["2", 3]); // Ok
    }

    static void Test(params MyCollection a)
    {
    }
}
```
发现它的“OK”其实不OK啊，我用的是.net 8，确实对应csharp-12啊

**Reply by KiyanYang** [reply]
是的，`params MyCollection` 还没有正式发布，提案在这 https://github.com/dotnet/csharplang/issues/7700 ，可以先通过 [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUoKgW2QYQHsAbYiAYxgEtCA7ZAGhhATloB8ABABgAJOAjADoipCtTrIA3AFgAUD34CALLLnzOAJl4BZAJ6iylGrV4glAZnkBveb3u8AkgFFacHBGwxCCS0IDiEDCu7p6wPgAUAJS8ALwAfLwwABYIhADuvG6kanYOnBZOIR5ePgA8gtyJgcFuJeEI0XGJKWmZ2cRSvHn2BfzKvACCACbDEYTAAFbivMQxCUmpGVlwOfIAvvIa2oIA7DY9SgBs/bpQVLTRh7ZyDne8ACooMBEARJqvDLwWUV0A9H9eJ40r58ABlAQAVm4FnMgwQAHNQrQYLxNOZyFBaLRCKjyHQAG6eVFgNI4XgAcguMApSUIlMqFMOdyeyBeAG13p9vgBdX68AG8ADyAGtDpt1Ld8gITpwBqyXgAHLC4ZC6AwkIwSUxQKLXcUbIA==)预览一下

**Reply by tsctsc6** [reply]
原来如此，谢谢
